.travis.yml

```diff
@@ -9,12 +9,11 @@ matrix:
 
 env:
   global:
-  - COMPOSER_BIN_DIR=$(composer global config bin-dir --absolute)
+    - COMPOSER_BIN_DIR=$(composer global config bin-dir --absolute)
 
 before_install:
-- composer install
-- composer global require friendsofphp/php-cs-fixer:^2
-- composer global require phpunit/phpunit
```

Comment:
For "external" tools like PHP-CS-Fixer it's fine because the tool doesn't load any of your code so there can't be any namespace conflict. It can just be annoying as a contributor but at least it doesn't pollute the project namespace. There is alternatives like https://github.com/bamarni/composer-bin-plugin to help out but that can be done in a separate PR.

However for PHPUnit it's a different story, not having it as a dependency means:

- Not being able to typehint PHPUnit API while writing tests
- Break any SA tool (because you are depending on classes that don't exist)
- Make it hard for contributors (require them to install PHPUnit separately, either a PHAR or globally but then it might be an issue when you have multiple projects)
- Least but not least: IMO it's a project dependency not a global one. Your project extends its API (meaning you depend on a specific PHPUnit version) and PHPUnit executes your code which means there might be namespace conflicts.

So for PHPUnit it's better to have it as a dev dependency and all the problems above are solved

Reply:
Thanks for the super detailed explanation here, super helpful!